### PR TITLE
Implementation of noise calibration for MFT

### PR DESCRIPTION
**Pull request by mcoquet642**
Implements a new DPL for MFT calibration, named `o2-calibration-mft-calib-workflow`. It takes as an input either decoded digits (if one specifies the `--useDigits` option) or clusters (default) and outputs CCDB payloads containing four different `o2::itsmft::NoiseMap` objects to be sent to the `ccdb-populator` device. This DPL is meant to be ran in a workflow command on a FLP, and the four noise maps correpond to the four different half faces processed by this FLP, which are written in different subfolders in the CCDB.

### CODE REVIEW

File: Detectors/ITSMFT/MFT/calibration/src/NoiseCalibrator.cxx [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+      int half = chipMap[id].half;
+      int layer = chipMap[id].layer;
+      int disk = chipMap[id].disk;
+      int face = layer % 2;
+
+      // Fast 1-pixel calibration
+      if ((rowSpan == 1) && (colSpan == 1)) {
+        if (half == 0 && face == 0) {
+          mNoiseMapH0F0.increaseNoiseCount(id, row, col);
+          mPath[0] = "/MFT/Calib/NoiseMap/h" + std::to_string(half) + "-d" + std::to_string(disk) + "-f" + std::to_string(face);
```
Comment (shahor02) [inline Detectors/ITSMFT/MFT/calibration/src/NoiseCalibrator.cxx]: why do you need to set the path after processing every digit?

File: Detectors/ITSMFT/MFT/calibration/src/NoiseCalibrator.cxx [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+      // All-pixel calibration
+      auto nBits = rowSpan * colSpan;
+      int ic = 0, ir = 0;
+      for (unsigned int i = 2; i < patt.getUsedBytes() + 2; i++) {
+        unsigned char tempChar = patt.getByte(i);
+        int s = 128; // 0b10000000
+        while (s > 0) {
+          if ((tempChar & s) != 0) {
+            mNoiseMap.increaseNoiseCount(id, row + ir, col + ic);
```
Comment (shahor02) [inline Detectors/ITSMFT/MFT/calibration/src/NoiseCalibrator.cxx]: why in "all-pixel" calibration you increment both common and partial map, while in 1-pixel version only partial maps?

File: Detectors/ITSMFT/MFT/calibration/src/NoiseCalibrator.cxx [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    }
+  }
+  mNumberOfStrobes += rofs.size();
+  return (mNumberOfStrobes * mProbabilityThreshold >= mThreshold) ? true : false;
+}
+
+void NoiseCalibrator::finalize()
+{
+  LOG(INFO) << "Number of processed strobes is " << mNumberOfStrobes;
+  mNoiseMapH0F0.applyProbThreshold(mProbabilityThreshold, mNumberOfStrobes);
```
Comment (shahor02) [inline Detectors/ITSMFT/MFT/calibration/src/NoiseCalibrator.cxx:148]: what about mNoiseMap?

File: Detectors/ITSMFT/MFT/calibration/src/NoiseCalibratorDigitsSpec.cxx [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  std::stringstream ss(src);
+  std::string token;
+  std::vector<std::string> tokens;
+
+  while (std::getline(ss, token, delim)) {
+    token = (trim ? trimSpaceStream(token) : token);
+    if (!token.empty()) {
+      tokens.push_back(std::move(token));
+    }
+  }
```
Comment (shahor02) [inline Detectors/ITSMFT/MFT/calibration/src/NoiseCalibratorDigitsSpec.cxx:59]: this code getting replicated in different sources over and over... OK for now but I will add it to StringUtils then change it everywhere.

**Review by shahor02 — COMMENTED** [review]
@mcoquet642 I've added some comments below, but the general one: why do you need to create 4 different CCDB objects, each one containing all chips? How are you going to use such maps?
And what is the reason to run on the FLP, which leads to such a fragmentation? 
Why not adapting the ITS noise time-slots based calibration workflow, which runs on EPN and produces one single file? Alternatively, you can produce partial results on the FLPs then merge them on the aggregation server.

**Comment by mcoquet642** [comment]
@shahor02 Thanks for the comments. Each of the noise maps correspond to a specific part of the detector processed by a given FLP. We wanted to try this implementation to keep modularity on these maps, which we would like to be adapted to the detector architecture. In particular we might want to update the noisy pixels list only for specific parts of the detector. With only one file, it seems more difficult to refresh partially the list without having to run a complete acquisition. In principle the noise occupancy scan should be light enough to run on FLPs, and this fragmentation of the noise map would allow us to run either on FLP or EPN, and so would give us flexibility on how to define the best strategy for calibration.

File: Detectors/ITSMFT/MFT/calibration/src/NoiseCalibrator.cxx (hunk 1 above)
Comment (mcoquet642) [reply Detectors/ITSMFT/MFT/calibration/src/NoiseCalibrator.cxx]: Indeed it is not needed. I tried to fix this in the new commit.

File: Detectors/ITSMFT/MFT/calibration/src/NoiseCalibrator.cxx (hunk 2 above)
Comment (mcoquet642) [reply Detectors/ITSMFT/MFT/calibration/src/NoiseCalibrator.cxx]: This is a mistake. The mNoiseMap object should not be there any more. I deleted it in the new commit.

File: Detectors/ITSMFT/MFT/calibration/src/NoiseCalibrator.cxx (hunk 3 above)
Comment (mcoquet642) [reply Detectors/ITSMFT/MFT/calibration/src/NoiseCalibrator.cxx:148]: Same as above.

File: Detectors/ITSMFT/MFT/calibration/src/NoiseCalibratorDigitsSpec.cxx (hunk 4 above)
Comment (mcoquet642) [reply Detectors/ITSMFT/MFT/calibration/src/NoiseCalibratorDigitsSpec.cxx:59]: Yes indeed I saw that it was already present in other sources ... Adding in to StringUtils would be good.

Comment (shahor02) [reply Detectors/ITSMFT/MFT/calibration/src/NoiseCalibratorDigitsSpec.cxx:59]: Now there is this method: https://github.com/AliceO2Group/AliceO2/blob/73ac08987c77e530cca2b1b8796d7180404f4cb4/Common/Utils/include/CommonUtils/StringUtils.h#L99

**Review by shahor02 — COMMENTED** [review]
Hi @mcoquet642

But each of your 4 maps cover all chips. And the current application of the noise map assumes just 1 map. 
You are currently splitting the maps per half and face, so each map covers more or less uniformly all layers, what would be the reason of updating the map only for e.g. even layers? This would also complicate automatic operation.
Also, in this way you are coupling your the calibration schema to particular detector-FLP cabling, what if one day recable it?

This workflow cannot run on the EPN in standard data routing mode, it would require necessarily pushing the data from all FLPs to the same EPN.

**Comment by mcoquet642** [comment]
@shahor02 Indeed it seems that to correctly register the chip ID in the NoiseMap objects, they must be declared to cover all chips. In principle we would have liked to use maps of only 68 chips (the number of chips on the largest half face).

I didn't understand what you wrote about layers. Each of the 5 FLPs treats four different half faces, two of them on a half-disk of h0 and the other two on another half-disk of h1 (e.g. h0-d0-f0,1 and h1-d4-f0,1). Running the workflow on a given FLP and a noise acquisition only on a particular half face processed by this FLP face would enable us to update only this half face without touching anything else.

**Comment by shahor02** [comment]
> @shahor02 Indeed it seems that to correctly register the chip ID in the NoiseMap objects, they must be declared to cover all chips. In principle we would have liked to use maps of only 68 chips (the number of chips on the largest half face).

Currently both ITS and MFT are noise masking uses single array with contiguous chip ID, so the mask is directly queried by the chipID. How are you going to use 4 maps? Unless you merge 4 maps to 1, you will need to branch out the MFT version to have operate with 4 different maps, add to the query switching from one map to another for given chip half/face when calling `AlpideCoder::decodeChip`. Not sure it is worth it.
 
> I didn't understand what you wrote about layers. Each of the 5 FLPs treats four different half faces, two of them on a half-disk of h0 and the other two on another half-disk of h1 (e.g. h0-d0-f0,1 and h1-d4-f0,1). Running the workflow on a given FLP and a noise acquisition only on a particular half face processed by this FLP face would enable us to update only this half face without touching anything else.

No, you cannot calibrate reading out only e.g. 1 half-face of single FLP, since you anyway create an object for 5 half-faces: then you will upload to the ccdb an object with other 3 half-faces of this map empty and this object will override the calibration of these half-faces from previous object.
The smallest granularity you can afford with the code like that is to run the acquisition for all 5 half-faces contributing to one of your map, and I am trying to imagine the situation where you will need this.

BTW, in sendOutput you are sending to ccdb all 4 maps.

**Comment by arakotoz** [comment]
Hi @shahor02 
> This workflow cannot run on the EPN in standard data routing mode, it would require necessarily pushing the data from all FLPs to the same EPN.

This is indeed our intention. Since we plan to run the threshold scan DPL in this way (and it is a DPL much more complex w.r.t. the noise scan DPL), testing the noise scan DPL in this data distribution scheme is a very good opportunity/iteration towards a successful use of data distribution scheme based on FEEid.

**Comment by shahor02** [comment]
@arakotoz 

For the threshold scan the data of given FLP should be pushed to the given EPN, but different FLPs will use different EPNs. Otherwise there I don't see a point of pushing it on the EPN.

**Comment by arakotoz** [comment]
Hi @shahor02,
> No, you cannot calibrate reading out only e.g. 1 half-face of single FLP, since you anyway create an object for 5 half-faces: then you will upload to the ccdb an object with other 3 half-faces of this map empty and this object will override the calibration of these half-faces from previous object.
> The smallest granularity you can afford with the code like that is to run the acquisition for all 5 half-faces contributing to one of your map, and I am trying to imagine the situation where you will need this.

Running the acquisition with only one FLP (and the detector parts connected to that FLP) will create the noise map for 4 half-faces. For e.g., if we use FLP182, we would be able to read the following parts of the detector: h0-d0-f0, h0-d0-f1, h1-d4-f0 and h1-d4-f1 (half MFT, disk number, face of the half disk). But it is indeed annoying that if we disable 3 of them (for e.g. we would only recalibrate h0-d0-f0) then the noise maps for the 3 disabled parts still end up in the CCDB.

**Comment by arakotoz** [comment]
> @arakotoz
> 
> For the threshold scan the data of given FLP should be pushed to the given EPN, but different FLPs will use different EPNs. Otherwise there I don't see a point of pushing it on the EPN.

The idea for the noise scan is to have 1 EPN for each half-face. If I take again the example of the FLP182, that would mean the data read for h0-d0-f0 would be sent on EPN-0, the data read for h0-d0-f1 would be sent to EPN-1, the data read for h1-d4-f0 would be sent to EPN-2 and the data read for h1-d4-f1 would be sent to EPN-3. So for each FLP, the data distribution scheme based on FEEid would send the data to 4 different EPNs. We have 5 FLPs, so we would use 20 EPNs.

We could start with the same data distribution scheme for the threshold scan.

**Comment by arakotoz** [comment]
@shahor02 

> Currently both ITS and MFT are noise masking uses single array with contiguous chip ID, so the mask is directly queried by the chipID. How are you going to use 4 maps? Unless you merge 4 maps to 1, you will need to branch out the MFT version to have operate with 4 different maps, add to the query switching from one map to another for given chip half/face when calling `AlpideCoder::decodeChip`. Not sure it is worth it.

I guess you are referring here to masking done **after** readout, as opposed to the masking we can do before readout, i.e. during the sensor configuration by the DCS, which prevent the noisy pixels to send any data to readout at all. The masking done before readout does not know anything about the flat software chip index anyway, so the noise map are not useable as such without any transformation to get back the hardware chip id that is understandable by the DCS and the RUs.

For the masking done by O2 after readout: 
- how is it invoked in the workflow raw data -> decoding (+digits) + clustering -> reconstruction? is it already enabled by default during digits or clustering production? how useful is it if we already mask the noisy pixels before readout? it is there in case new noisy pixels appears during the long runs in the current fill and we don't have the opportunity yet to refresh the noise calibration? it it there in case the sensors "forget" their masking map during the long runs? 
- yes, the minimal course of action would be to add a small algorithmic layer between the current query implementation and the masking. This layer would take in charge the dynamic "merging" that you were referring to. And the query would go to it, instead of directly to the CCDB.

**Comment by shahor02** [comment]
@arakotoz let's keep discussion either here or via email, but not both at the same time... On top of what I just wrote in the email:

> For e.g., if we use FLP182, we would be able to read the following parts of the detector: h0-d0-f0, h0-d0-f1, h1-d4-f0 and h1-d4-f1 (half MFT, disk number, face of the half disk). But it is indeed annoying that if we disable 3 of them (for e.g. we would only recalibrate h0-d0-f0) then the noise maps for the 3 disabled parts still end up in the CCDB.

This should be forbidden anyway, since 3 dummy maps will hide the calibration you wanted to keep.

> The idea for the noise scan is to have 1 EPN for each half-face. If I take again the example of the FLP182, that would mean the data read for h0-d0-f0 would be sent on EPN-0, the data read for h0-d0-f1 would be sent to EPN-1, the data read for h1-d4-f0 would be sent to EPN-2 and the data read for h1-d4-f1 would be sent to EPN-3. So for each FLP, the data distribution scheme based on FEEid would send the data to 4 different EPNs. We have 5 FLPs, so we would use 20 EPNs.
> We could start with the same data distribution scheme for the threshold scan.

But the threshold scan you are always running for the full detector, right, i.e. 1 special configuration would be enough. 
In this case, the configuration will depend on which half-faces you are going to re-calibrate, so it is not really the same configuration.

> I guess you are referring here to masking done after readout, as opposed to the masking we can do before readout, i.e. during the sensor configuration by the DCS, which prevent the noisy pixels to send any data to readout at all. The masking done before readout does not know anything about the flat software chip index anyway, so the noise map are not useable as such without any transformation to get back the hardware chip id that is understandable by the DCS and the RUs.

The noise map you are using comes from the ITS which does use software chip numbering. And it is used exactly in the software noise suppression, see my answer by email.

> For the masking done by O2 after readout:
>   * how is it invoked in the workflow raw data -> decoding (+digits) + clustering -> reconstruction? is it already enabled by default during digits or clustering production? how useful is it if we already mask the noisy pixels before readout? it is there in case new noisy pixels appears during the long runs in the current fill and we don't have the opportunity yet to refresh the noise calibration? it it there in case the sensors "forget" their masking map during the long runs?

It is supposed to be loaded from the CCDB (currently is read from file if available, see `--noise-file` option processing in `o2-itsmft-stf-decoder-workflow`. As I wrote before, even if HW suppression is enough for raw data processing, the map has to be applied in MC.

>  *  yes, the minimal course of action would be to add a small algorithmic layer between the current query implementation and the masking. This layer would take in charge the dynamic "merging" that you were referring to. And the query would go to it, instead of directly to the CCDB.

So, what is the point of creating 20 objects in a complicated way requiring special configuration and then merge it to 1 object, instead of using already existing workflow which produces 1 object ?

**Pull request closed without merging**